### PR TITLE
PADV-351: Add LTI XBlock and Courseware views

### PR DESCRIPTION
**Pull request by kuipumu**
## Ticket

https://agile-jira.pearson.com/browse/PADV-351

## Description

This PR adds XBlock and Courseware views for course LTI launches, this will allow us to have our own endpoints to render the course content for LTI launches with only the unit content. To achieve this we used the render_xblock function from edx-platform.

## Type of Change

- [x] Add LTI XBlock and Courseware views
- [x] Add courseware.html template.
- [x] Add edxapp_wrapper module.
- [x] Add courseware module wrapper backend.
- [x] Add OLTITP_COURSEWARE_BACKEND setting.
- [x] Add required edx-platform test settings.
- [x] Imported render_xblock from edx-platform.
- [x] Modify LtiToolLaunchView post method params.
- [x] Remove usage_key from LtiToolLaunchView.
- [x] Update LtiToolLaunchView tests.
- [x] Include LtiXBlockView and LtiCoursewareView tests
- [x] Update urls module tests.

## Testing:

- Run the LMS: `make dev.up.lms`.
- Install `openedx_lti_tool_plugin` on the LMS.
- Add required settings to LMS.
	
```
    # Enable LTI Tool Provider Plugin.
    OLTITP_ENABLE_LTI_TOOL = True
```
- Go to http://localhost:18000/openedx_lti_tool_plugin/course/unit_usage_key (Replace unit_usage_key with a XBlock ID of course unit).
- The browser should display the requested content on an iframe using the courseware.html template.

## Note

Currently the template only displays the content, we still need to render the course outline to let the user navigate the course, we also need to add the course home to make LTI 1.3 launches redirect the users to it.

The XBlock and Courseware views are only able to display units properly, other types of XBlock either don't render or render a broken navigation menu.

## Reviewers

- [x] @Squirrel18 
- [x] @Jacatove 
- [x] @alexjmpb 
- [x] @anfbermudezme

### CODE REVIEW

**Comment by anfbermudezme** [comment]
Could you please add more context, how to configure the course correctly to see the LTI?

image

Also the setting `FEATURES['ENABLE_LTI_PROVIDER_PLUGIN'] = True`, shouldn't it be `FEATURES['OLTTP_ENABLE_LTI_TOOL'] = True`?

**Comment by kuipumu** [comment]
@anfbermudezme I updated the PR letter, the correct settings should be OLTTP_ENABLE_LTI_TOOL = True, You only need to get a unit ID from any course and sent that ID to the courseware URL: http://localhost:18000/openedx_lti_tool_plugin/course/block-v1:edX+DemoX+Demo_Course+type@vertical+block@vertical_0270f6de40fc

A sequence and component ID can also render an XBlock, in the case of a sequence, it will render a broken navigation menu on top, I'm still working out if we could fix that or render the course outline using our own template.

File: openedx_lti_tool_plugin/views.py [hunk 1]
```diff
@@ -197,6 +206,64 @@ def get(self, request: HttpRequest) -> JsonResponse:
...
+        Returns an XBlock using render_xblock view.
+
+        Args:
+            request: HTTP request object.
+            usage_key_string: Usage key string.
+
+        Returns:
+            HTTP response with rendered LTI courseware view.
+        """
+        # Currently, render_xblock calls is_learning_mfe to evaluate if the request
```
Comment (anfbermudezme) [inline openedx_lti_tool_plugin/views.py:233]: What do you think about reduce this inline comment?

**Review by anfbermudezme — CHANGES_REQUESTED** [review]

File: openedx_lti_tool_plugin/settings/common.py [hunk 2]
```diff
@@ -38,3 +38,6 @@ def plugin_settings(settings: LazySettings):
     """
     settings.OLTTP_ENABLE_LTI_TOOL = False
     settings.AUTHENTICATION_BACKENDS.append('openedx_lti_tool_plugin.auth.LtiAuthenticationBackend')
+
+    # Backends settings
+    settings.OLTTP_COURSEWARE_BACKEND = 'openedx_lti_tool_plugin.edxapp_wrapper.backends.courseware_module_o_v1'
```
Comment (Squirrel18) [inline openedx_lti_tool_plugin/settings/common.py:43]:
```suggestion
    settings.OLTITP_COURSEWARE_BACKEND = 'openedx_lti_tool_plugin.edxapp_wrapper.backends.courseware_module_o_v1'
```
Shouldn't it be OLTITP (Open edX LTI Tool provider)? @kuipumu

File: openedx_lti_tool_plugin/templates/openedx_lti_tool_plugin/courseware.html [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    <meta name='viewport' content='width=device-width, initial-scale=1'>
+  </head>
+
+  <body>
+    <iframe
+      id='unit-iframe'
+      src='{{ xblock_url }}'
+      allow='microphone *; camera *; midi *; geolocation *; encrypted-media *'
+      allowfullscreen=''
+      scrolling='no'>
```
Comment (Squirrel18) [inline openedx_lti_tool_plugin/templates/openedx_lti_tool_plugin/courseware.html:14]:
```suggestion
      scrolling='no'
```

File: openedx_lti_tool_plugin/templates/openedx_lti_tool_plugin/courseware.html [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  <head>
+    <meta charset='utf-8'>
+    <meta name='viewport' content='width=device-width, initial-scale=1'>
+  </head>
+
+  <body>
+    <iframe
+      id='unit-iframe'
+      src='{{ xblock_url }}'
+      allow='microphone *; camera *; midi *; geolocation *; encrypted-media *'
```
Comment (Squirrel18) [inline openedx_lti_tool_plugin/templates/openedx_lti_tool_plugin/courseware.html:12]: Do you think that is necessary to set these values?

File: openedx_lti_tool_plugin/templates/openedx_lti_tool_plugin/courseware.html [hunk 5]
```diff
@@ -0,0 +1,37 @@
+<!doctype html>
+<html>
+  <head>
+    <meta charset='utf-8'>
+    <meta name='viewport' content='width=device-width, initial-scale=1'>
+  </head>
+
```
Comment (Squirrel18) [inline openedx_lti_tool_plugin/templates/openedx_lti_tool_plugin/courseware.html:7]: Unnecessary blank line.

File: openedx_lti_tool_plugin/templates/openedx_lti_tool_plugin/courseware.html [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  <body>
+    <iframe
+      id='unit-iframe'
+      src='{{ xblock_url }}'
+      allow='microphone *; camera *; midi *; geolocation *; encrypted-media *'
+      allowfullscreen=''
```
Comment (Squirrel18) [inline openedx_lti_tool_plugin/templates/openedx_lti_tool_plugin/courseware.html:13]: This content is intended to be embedded on other platforms, so I don't think it is necessary to allow this attribute. @kuipumu

File: openedx_lti_tool_plugin/templates/openedx_lti_tool_plugin/courseware.html [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    </iframe>
+    <style>
+      #unit-iframe {
+          width: 100%;
+          border: none;
+          display: block;
+      }
+    </style>
+    <script>
+      // Detect message events from the XBlock iframe.
```
Comment (Squirrel18) [inline openedx_lti_tool_plugin/templates/openedx_lti_tool_plugin/courseware.html:24]: Why is this required? @kuipumu

**Review by Squirrel18 — CHANGES_REQUESTED** [review]

File: openedx_lti_tool_plugin/templates/openedx_lti_tool_plugin/courseware.html (hunk 3 above)
Comment (kuipumu) [reply openedx_lti_tool_plugin/templates/openedx_lti_tool_plugin/courseware.html:14]: Do you mean remove scrolling or the closing tag?, removing the tag breaks the iframe

File: openedx_lti_tool_plugin/templates/openedx_lti_tool_plugin/courseware.html (hunk 4 above)
Comment (kuipumu) [reply openedx_lti_tool_plugin/templates/openedx_lti_tool_plugin/courseware.html:12]: @Squirrel18 They allow XBlocks to be able to interact with the parent window. I also added fullscreen since allowfullscreen is deprecated.

File: openedx_lti_tool_plugin/templates/openedx_lti_tool_plugin/courseware.html (hunk 7 above)
Comment (kuipumu) [reply openedx_lti_tool_plugin/templates/openedx_lti_tool_plugin/courseware.html:24]: This receives the Window.postMessage() sent on the render_xblock template view lms/templates/courseware/courseware-chromeless.html, with this we can receive the plugin.resize message to get the height and width reported by the iframe, with this we can have a responsive height by making the parent window change the iframe size dynamically according to the iframe reported size.

**Review by anfbermudezme — APPROVED** [review]

File: openedx_lti_tool_plugin/views.py [hunk 8]
```diff
@@ -197,6 +206,65 @@ def get(self, request: HttpRequest) -> JsonResponse:
...
+            unit_key: Unit key string.
+
+        Returns:
+            HTTP response with rendered LTI courseware view.
+        """
+        return self.render_to_response(
+            {
+                'lms_root_url': settings.LMS_ROOT_URL,
+                'xblock_url': reverse(f'{AppConfig.name}:lti-xblock', args=[unit_key]),
+            }
```
Comment (Squirrel18) [inline openedx_lti_tool_plugin/views.py:269]: Trailing comma. @kuipumu

File: openedx_lti_tool_plugin/views.py [hunk 9]
```diff
@@ -197,6 +206,65 @@ def get(self, request: HttpRequest) -> JsonResponse:
...
+        Returns:
+            HTTP response with rendered LTI courseware view.
+        """
+        # render_xblock calls is_learning_mfe to evaluate if the request
+        # is sent by the learning MFE by evaluating the HTTP Referer header,
+        # this enables some optimizations to make the view work properly
+        # on an iframe, for now, we will modify the request meta to pretend
+        # its from the learning MFE URL until render_xblock is improved
+        # to enable this optimizations from other addresses.
+        request.META['HTTP_REFERER'] = settings.LEARNING_MICROFRONTEND_URL
```
Comment (Squirrel18) [inline openedx_lti_tool_plugin/views.py:239]: Add a safe getter for LEARNING_MICROFRONTEND_URL. @kuipumu

File: openedx_lti_tool_plugin/views.py [hunk 10]
```diff
@@ -197,6 +206,65 @@ def get(self, request: HttpRequest) -> JsonResponse:
...
+        Args:
+            request: HTTP request object.
+            unit_key: Unit key string.
+
+        Returns:
+            HTTP response with rendered LTI courseware view.
+        """
+        return self.render_to_response(
+            {
+                'lms_root_url': settings.LMS_ROOT_URL,
```
Comment (Squirrel18) [inline openedx_lti_tool_plugin/views.py:267]: Add a safe getter for LMS_ROOT_URL. @kuipumu

File: openedx_lti_tool_plugin/tests/__init__.py [hunk 11]
```diff
@@ -3,3 +3,6 @@
 ISS = 'http://foo.example.com/'
 SUB = 'random-sub'
 AUD = 'random.aud.app'
+
+COURSE_ID = 'course-v1:test+test+test'
+UNIT_KEY = 'block-v1:test+test+test+type@test@0'
```
Comment (Squirrel18) [inline openedx_lti_tool_plugin/tests/__init__.py:8]: The correct name is Usage Key, so please change it to USAGE_KEY. @kuipumu

File: openedx_lti_tool_plugin/tests/test_views.py [hunk 12]
```diff
@@ -383,7 +377,75 @@ def test_get_jwks_mock(self, get_jwks_mock: MagicMock):
...
+        render_xblock_mock.assert_called_once_with(request, UNIT_KEY, check_if_enrolled=False)
+
+    @override_settings(OLTITP_ENABLE_LTI_TOOL=False)
+    def test_with_lti_disabled(self):
+        """Test raise 404 response when plugin is disabled."""
+        with self.assertRaises(Http404):
+            self.view_class.as_view()(self.factory.get(self.url))
+
+
+class TestLtiCoursewareView(FactoryMixin, TestCase):
```
Comment (Squirrel18) [inline openedx_lti_tool_plugin/tests/test_views.py:416]: Shouldn't it also return 404 in case LTI is not active? @kuipumu

**Review by Squirrel18 — CHANGES_REQUESTED** [review]

File: openedx_lti_tool_plugin/tests/test_views.py (hunk 12 above)
Comment (kuipumu) [reply openedx_lti_tool_plugin/tests/test_views.py:416]: @kuipumu There is already a test_with_lti_disabled on TestLtiCoursewareView to test this scenario

File: openedx_lti_tool_plugin/tests/test_views.py [hunk 13]
```diff
@@ -383,7 +376,75 @@ def test_get_jwks_mock(self, get_jwks_mock: MagicMock):
...
+        super().setUp()
+        self.url = reverse('lti-courseware', args=[USAGE_KEY])
+        self.view_class = LtiCoursewareView
+
+    @patch('openedx_lti_tool_plugin.views.reverse')
+    @patch('openedx_lti_tool_plugin.views.LtiCoursewareView.render_to_response')
+    def test_with_course_sequence_unit_ids(
+        self,
+        render_to_response_mock: MagicMock,
+        reverse_mock: MagicMock
```
Comment (Squirrel18) [inline openedx_lti_tool_plugin/tests/test_views.py:429]: Trailing comma. @kuipumu

File: openedx_lti_tool_plugin/tests/test_views.py [hunk 14]
```diff
@@ -383,7 +376,75 @@ def test_get_jwks_mock(self, get_jwks_mock: MagicMock):
...
+            reverse_mock: Mocked reverse function.
+        """
+        request = self.factory.get(self.url)
+        self.view_class.as_view()(request, USAGE_KEY)
+
+        render_to_response_mock.assert_called_once_with(
+            {
+                'lms_root_url': '',
+                'xblock_url': reverse_mock('lti-xblock', args=[USAGE_KEY]),
+            }
```
Comment (Squirrel18) [inline openedx_lti_tool_plugin/tests/test_views.py:444]: Trailing comma. @kuipumu

**Review by Squirrel18 — CHANGES_REQUESTED** [review]

**Review by Squirrel18 — APPROVED** [review]

File: openedx_lti_tool_plugin/tests/test_urls.py [hunk 15]
```diff
@@ -14,8 +21,22 @@ def test_lti_tool_login_url_resolves(self):
...
+            LtiToolLaunchView,
+        )
 
     def test_lti_tool_jwks_url_resolves(self):
         """Test LtiToolLoginView URL can be resolved."""
         self.assertEqual(resolve(reverse('lti1p3-pub-jwks')).func.view_class, LtiToolJwksView)
+
+    def test_lti_xblock_url_resolves(self):
+        """Test LtiXBlockView URL can be resolved."""
+        self.assertEqual(resolve(reverse('lti-xblock', args=[USAGE_KEY])).func.view_class, LtiXBlockView)
```
Comment (alexjmpb) [inline openedx_lti_tool_plugin/tests/test_urls.py:35]:
```suggestion
        self.assertEqual(
            resolve(reverse('lti-xblock', args=[USAGE_KEY])).func.view_class,
            LtiXBlockView,
        )
```

File: openedx_lti_tool_plugin/tests/test_urls.py [hunk 16]
```diff
@@ -14,8 +21,22 @@ def test_lti_tool_login_url_resolves(self):
...
         """Test LtiToolLaunchView URL can be resolved."""
-        self.assertEqual(resolve(reverse('lti1p3-launch')).func.view_class, LtiToolLaunchView)
+        self.assertEqual(
+            resolve(reverse('lti1p3-launch', args=[COURSE_ID])).func.view_class,
+            LtiToolLaunchView,
+        )
 
     def test_lti_tool_jwks_url_resolves(self):
         """Test LtiToolLoginView URL can be resolved."""
         self.assertEqual(resolve(reverse('lti1p3-pub-jwks')).func.view_class, LtiToolJwksView)
```
Comment (alexjmpb) [inline openedx_lti_tool_plugin/tests/test_urls.py:31]:
```suggestion
        self.assertEqual(
            resolve(reverse('lti1p3-pub-jwks')).func.view_class,
            LtiToolJwksView
        )
```

File: openedx_lti_tool_plugin/urls.py [hunk 17]
```diff
@@ -1,10 +1,27 @@
...
 
 urlpatterns = [
-    path('1.3/login/', views.LtiToolLoginView.as_view(), name='lti1p3-login'),
-    path('1.3/launch/', views.LtiToolLaunchView.as_view(), name='lti1p3-launch'),
-    path('1.3/pub/jwks/', views.LtiToolJwksView.as_view(), name='lti1p3-pub-jwks'),
+    path('1.3/', include([
+        path('login', views.LtiToolLoginView.as_view(), name='lti1p3-login'),
+        path('pub/jwks', views.LtiToolJwksView.as_view(), name='lti1p3-pub-jwks'),
+        re_path(
+            fr'^launch/{settings.COURSE_ID_PATTERN}$',
```
Comment (alexjmpb) [inline openedx_lti_tool_plugin/urls.py:12]: Can't this be imported from the lms?

Comment (kuipumu) [reply openedx_lti_tool_plugin/urls.py:12]: @Squirrel18 I will already use the setting from the LMS while installed, on the cli it will use the test settings, I could patch it on tests and not add it to the settings, so we don't need to keep a copy for tests.

**Review by alexjmpb — APPROVED** [review]

**Review by Jacatove — CHANGES_REQUESTED** [review]

File: README.rst [hunk 18]
```diff
@@ -1,27 +1,53 @@
...
 
-  # Set up a virtualenv
+- Install plugin on your server (in this case the devstack docker lms shell):
+
+.. code-block:: bash
+
+  cd ~openedx/devstack/  # Change for your devstack path (if you are using devstack)
+  make lms-shell  # Shell into the lms container (or server where lms process runs)
+  pip install -e /edx/src/openedx-lti-tool-plugin
+  /edx/app/edxapp/edx-platform/manage.py lms migrate  # Run plugin migrations
```
[the remark this answers is not in the document]
Comment (kuipumu) [reply README.rst:28]: @Jacatove Could you please elaborate further?, I didn't understood the suggestion

**Review by Jacatove — APPROVED** [review]
